File: backend/src/main/java/com/example/socinet/service/AdminService.java
package com.example.socinet.service;

import com.example.socinet.dto.AccountDto;
import com.example.socinet.dto.PostDto;
import com.example.socinet.entity.Account;
import com.example.socinet.entity.Post;
import com.example.socinet.repository.AccountRepository;
import com.example.socinet.repository.PostRepository;
import com.example.socinet.repository.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class AdminService {
    private final PostRepository postRepo;
    private final AccountRepository accountRepo;

    public PostDto managePost(Long postId, String action) throws Exception {
        Optional<Post> postOpt = postRepo.findById(postId);
        if(postOpt.isEmpty()) throw new Exception("Post not found");
        switch (action){
            case "activate":
                // update isActive = true
                postOpt.get().setActive(true);
                break;
            case "inactivate":
                // update isActive = false
                postOpt.get().setActive(false);
                break;
            default: break;
        }
        return new PostDto(postRepo.save(postOpt.get()));
    }

    public long getNumberOfPostByUserId(Long userId){
        long count = postRepo.countByUser_Id(userId);
        return count;
    }

    public List<AccountDto> getAccountsByName(String name, Pageable pageable){
        List<Account> accounts = accountRepo.findByUser_NameContainingIgnoreCase(name, pageable);
        List<AccountDto> accountsDto = new ArrayList<>();
        accounts.forEach(account -> accountsDto.add(new AccountDto(account)));
        return accountsDto;
    }

    public AccountDto getAccount(String username) throws Exception{
        Account account = accountRepo.findAccountByUsername(username).orElseThrow(() -> new Exception("Account is not exist"));
        return new AccountDto(account);
    }

    public long getNumberOfAccountByName(String name){
        long count = accountRepo.countByUser_NameContainingIgnoreCase(name);
        return count;
    }

    public AccountDto manageAccount(String username, String action) throws Exception {
        Optional<Account> accountOpt = accountRepo.findAccountByUsername(username);
        if(accountOpt.isEmpty()) throw new Exception("Account not found");

        switch(action){
            case "activate":
                accountOpt.get().setActive(true);
                break;
            case "inactivate":
                accountOpt.get().setActive(false);
            default: break;
        }

        return new AccountDto(accountRepo.save(accountOpt.get()));
    }

}

File: frontend/src/api/AdminService.js
import axios from "axios";

const BASE = "api/v1/admin";

const AdminService = {
  getAccountByName: async (name, page = 0, size = 5) =>
    axios
      .get(BASE + "/accounts", { params: { name, page, size } })
      .then((res) => res.data)
      .catch((e) => e.response.data),

  countNumberOfAccountByName: async (name) =>
    axios
      .get(BASE + "/count/accounts", { params: { name } })
      .then((res) => res.data)
      .catch((e) => e.response.data),

  getPostByUserId: async (userId, page = 0, size = 5) =>
    axios
      .get(BASE + "/posts", { params: { userId, page, size } })
      .then((res) => res.data)
      .catch((e) => e.response.data),

  countNumberOfPostByUserId: async (userId) =>
    axios
      .get(BASE + "/count/posts", { params: { userId } })
      .then((res) => res.data)
      .catch((e) => e.response.data),
  manageAccount: async (username, isActive) =>
    axios
      .put(
        BASE + "/accounts/" + username,
        {},
        {
          params: { action: isActive ? "activate" : "inactivate" },
        }
      )
      .then((res) => res.data)
      .catch((e) => e.response.data),
  managePost: async (postId, isActive) =>
    axios
      .put(
        BASE + "/posts/" + postId,
        {},
        {
          params: { action: isActive ? "activate" : "inactivate" },
        }
      )
      .then((res) => res.data)
      .catch((e) => e.response.data),
};

export const {
  getAccountByName,
  getPostByUserId,
  countNumberOfAccountByName,
  countNumberOfPostByUserId,
  manageAccount,
  managePost,
} = AdminService;
